### PR TITLE
Optimize might_contain Spark function

### LINKED ISSUE #7363: Improve spark MightContain function performance
### Description

By below changes, we improve the performance of `MightContain` function:
1. Add `applyToSelected` to support batch applying: if all rows are selected, then we handle them in batch mode; if there are consecutive 64 rows which are selected (`kAllSet == word`), we also handle them in batch mode.
2. Implement MightContain by SIMD.

**TPCDS query16**
Plan Node: 
-- Project[expressions: (n2_7:INTEGER, murmur3hash(42,"n1_12")), (n2_8:INTEGER, "n1_8"), (n2_9:INTEGER, "n1_9"), (n2_10:INTEGER, "n1_10"), (n2_11:INTEGER, "n1_11"), (n2_12:BIGINT, "n1_12"), (n2_13:DECIMAL(7,2), "n1_13"), (n2_14:DECIMAL(7,2), "n1_14")] -> n2_7:INTEGER, n2_8:INTEGER, n2_9:INTEGER, n2_10:INTEGER, n2_11:INTEGER, n2_12:BIGINT, n2_13:DECIMAL(7,2), n2_14:DECIMAL(7,2)
  -- Project[expressions: (n1_8:INTEGER, "n0_0"), (n1_9:INTEGER, "n0_1"), (n1_10:INTEGER, "n0_2"), (n1_11:INTEGER, "n0_3"), (n1_12:BIGINT, "n0_4"), (n1_13:DECIMAL(7,2), "n0_5"), (n1_14:DECIMAL(7,2), "n0_6")] -> n1_8:INTEGER, n1_9:INTEGER, n1_10:INTEGER, n1_11:INTEGER, n1_12:BIGINT, n1_13:DECIMAL(7,2), n1_14:DECIMAL(7,2)
    -- TableScan[table: hive_table, range filters: [(cs_call_center_sk, Filter(IsNotNull, deterministic, null not allowed)), (cs_ship_addr_sk, Filter(IsNotNull, deterministic, null not allowed)), (cs_ship_date_sk, Filter(IsNotNull, deterministic, null not allowed))], remaining filter: (and(and(**might_contain**("AQB==",xxhash64(42,"cs_ship_addr_sk")),**might_contain**("AQQ==",xxhash64(42,"cs_call_center_sk"))),**might_contain**("AQg==",xxhash64(42,"cs_ship_date_sk"))))] -> n0_0:INTEGER, n0_1:INTEGER, n0_2:INTEGER, n0_3:INTEGER, n0_4:BIGINT, n0_5:DECIMAL(7,2), n0_6:DECIMAL(7,2), n0_7:INTEGER

Before
image

After
image

**TPCDS query93**
Plan Node: 
-- Project[expressions: (n2_5:INTEGER, murmur3hash(42,"n1_8","n1_6")), (n2_6:INTEGER, "n1_6"), (n2_7:INTEGER, "n1_7"), (n2_8:BIGINT, "n1_8"), (n2_9:INTEGER, "n1_9"), (n2_10:DECIMAL(7,2), "n1_10")] -> n2_5:INTEGER, n2_6:INTEGER, n2_7:INTEGER, n2_8:BIGINT, n2_9:INTEGER, n2_10:DECIMAL(7,2)
  -- Project[expressions: (n1_6:INTEGER, "n0_0"), (n1_7:INTEGER, "n0_1"), (n1_8:BIGINT, "n0_2"), (n1_9:INTEGER, "n0_3"), (n1_10:DECIMAL(7,2), "n0_4")] -> n1_6:INTEGER, n1_7:INTEGER, n1_8:BIGINT, n1_9:INTEGER, n1_10:DECIMAL(7,2)
    -- TableScan[table: hive_table, remaining filter: (and(**might_contain**("AQA==",xxhash64(42,"ss_ticket_number")),**might_contain**("AQA==",xxhash64(42,"ss_item_sk"))))] -> n0_0:INTEGER, n0_1:INTEGER, n0_2:BIGINT, n0_3:INTEGER, n0_4:DECIMAL(7,2), n0_5:INTEGER

Before
image

After
image

**TPCDS query95**
Plan Node: 
-- Project[expressions: (n2_6:INTEGER, murmur3hash(42,"n1_10")), (n2_7:INTEGER, "n1_7"), (n2_8:INTEGER, "n1_8"), (n2_9:INTEGER, "n1_9"), (n2_10:BIGINT, "n1_10"), (n2_11:DECIMAL(7,2), "n1_11"), (n2_12:DECIMAL(7,2), "n1_12")] -> n2_6:INTEGER, n2_7:INTEGER, n2_8:INTEGER, n2_9:INTEGER, n2_10:BIGINT, n2_11:DECIMAL(7,2), n2_12:DECIMAL(7,2)
  -- Project[expressions: (n1_7:INTEGER, "n0_0"), (n1_8:INTEGER, "n0_1"), (n1_9:INTEGER, "n0_2"), (n1_10:BIGINT, "n0_3"), (n1_11:DECIMAL(7,2), "n0_4"), (n1_12:DECIMAL(7,2), "n0_5")] -> n1_7:INTEGER, n1_8:INTEGER, n1_9:INTEGER, n1_10:BIGINT, n1_11:DECIMAL(7,2), n1_12:DECIMAL(7,2)
    -- TableScan[table: hive_table, range filters: [(ws_ship_addr_sk, Filter(IsNotNull, deterministic, null not allowed)), (ws_ship_date_sk, Filter(IsNotNull, deterministic, null not allowed)), (ws_web_site_sk, Filter(IsNotNull, deterministic, null not allowed))], remaining filter: (and(and(**might_contain**("AQB==",xxhash64(42,"ws_ship_addr_sk")),**might_contain**("AQQ==",xxhash64(42,"ws_web_site_sk"))),**might_contain**("AQg==",xxhash64(42,"ws_ship_date_sk"))))] -> n0_0:INTEGER, n0_1:INTEGER, n0_2:INTEGER, n0_3:BIGINT, n0_4:DECIMAL(7,2), n0_5:DECIMAL(7,2), n0_6:INTEGER

Before
image

After
image

### PR DESCRIPTION
**Pull request by zhli1142015**
1. Add selected rows batch handle function.
2. Implement might contain with SIMD.

Fixes #7363

### CODE REVIEW

**Comment by zhli1142015** [comment]
@jinchengchenghh , @mbasmanova , could you please help to take a look?
Thanks.

File: velox/common/base/BloomFilter.h [hunk 1]
```diff
@@ -129,6 +135,58 @@ class BloomFilter {
...
   }
 
+  inline static void testForBatch(
+      const uint64_t* FOLLY_NONNULL bloom,
+      int32_t bloomSize,
+      const uint64_t* hashCode,
+      bool* result,
+      int32_t size) {
+    VELOX_DCHECK_EQ(size, 64);
+#if XSIMD_WITH_AVX2
```
Comment (mbasmanova) [inline velox/common/base/BloomFilter.h:145]: Can this be written using xsimd API?

What are the gains? Would you include a benchmark that demonstrates the perf win?

CC: @Yuhta

Comment (zhli1142015) [reply velox/common/base/BloomFilter.h:145]: I'm not familiar with the xsimd APi, let me check.
As `might_contain` function are evlauted in remaining filters, we use remaining filter time to measure the perf gain, for details number please see #7363 . Let me also add some benchmark for this.
Thanks.

File: velox/common/base/BloomFilter.h [hunk 2]
```diff
@@ -129,6 +135,58 @@ class BloomFilter {
     return mask == (bloom[index] & mask);
   }
 
+  inline static void testForBatch(
+      const uint64_t* FOLLY_NONNULL bloom,
+      int32_t bloomSize,
+      const uint64_t* hashCode,
+      bool* result,
```
Comment (Yuhta) [inline velox/common/base/BloomFilter.h:142]: `result` here should be a `uint64_t*` bit mask, then you can directly convert the comparison result to bit mask using `toBitMask` in `SimdUtil.h` and write directly into the `FlatVector<bool>` without row wise conversion (which is quite expensive and can be one of the bottleneck after you simdizing it).

File: velox/common/base/BloomFilter.h (hunk 1 above)
Comment (Yuhta) [reply velox/common/base/BloomFilter.h:145]: It's very straightforward to rewrite this using xsimd, everything here is built in except `gather`, you can get it from `SimdUtil.h`

**Pull request closed without merging**